qiime2/sdk/actiongraph.py

```diff
@@ -0,0 +1,217 @@
+                    d = {}
+                    d['inputs'] = i
+                    d['params'] = p
+                    d['outputs'] = output
+                    d['non_req'] = nr
+                    results.append(d)
+            else:
+                d = {}
+                d['inputs'] = i
+                d['params'] = p
+                d['outputs'] = output
+                d['non_req'] = []
+                results.append(d)
+
+    return results
+
+
+def unravel_union(l):
+    """
+    Unravel Union node to get all permutations of types for each action
+
+    Parameters
+    ----------
+    list : list of Qiime2.types
+
+    Returns
+    -------
+    list of lists - list of permuations of types for each action
+
+    """
+
+    result = [l]
+    for i, x in enumerate(l):
+        if 'members' in x.__dict__ and len(x.__dict__['members']) > 0:
+            members = list(x.__dict__['members'])
+            temp = copy.deepcopy(result)
+
+            # update result with first element of types in member
+            for each_list in result:
+                each_list[i] = members[0]
+
+            # add in other permutations of types in member
+            for n in range(1, len(members)):
+                copy_result = copy.deepcopy(temp)
+                for each_list in copy_result:
+                    each_list[i] = members[n]
+                result += copy_result
+    return result
+
+
+def build_graph(sigs=[]):
+    """
+    Constructs a networkx graph with different semantic types
+    and actions as nodes
+
+    Parameters
+    ----------
+    sigs : list of strings
+
+    Returns
+    -------
+    nx.DiGraph - networkx graph connected based on all or specified methods
+
+    """
+
+    G = nx.DiGraph()
+    action_list = []
+
+    # get all actions or specifc actions if specified in sigs
+    pm = qiime2.sdk.PluginManager()
+    if not sigs:
+        for pgn, pg in pm.plugins.items():
+            action_list += list(pg.actions.values())
+    else:
+        for pgn, pg in pm.plugins.items():
+            for action in pg.actions.keys():
+                if str(action) in sigs:
+                    action_list.append(pg.actions[str(action)])
+
+    for action in action_list:
+        node_combs = generate_nodes_by_action(action)
+
+        # each action node is a str(dict)
+        # each Qiime2.type node is the type itself
+        for dict in node_combs:
+            if not G.has_node(str(dict)):
+                G.add_node(str(dict), value=action)
+
+            for k in dict.keys():
+                for type in dict[k]:
+                    if not G.has_node(type):
+                        G.add_node(type, value=type)
+                    if k == 'outputs':
+                        G.add_edge(str(dict), type)
+                    elif k == 'inputs':
+                        G.add_edge(type, str(dict))
+                    elif k == 'params':
+                        G.add_edge(type, str(dict))
```

Comment:
**Change Requested**

A useful thing to store on the edge would be `name` which is the parameter name of that particular argument type. This is important because some actions have two parameters (cs definition os param) with the same type and it would be nice to be able to grab all incoming edges on a node and to have the parameter names that should be used.

Reply:
This change will require your `dict[k]` to itself be a dictionary, as a list certainly loses this context.